wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,632 @@
+		}
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, keyTypeString))
+	}
+
+	if err := c.client.FindObjectsInit(session, template); err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+	obj, _, err := c.client.FindObjects(session, 2)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjects: %s", err)
+	}
+	if err := c.client.FindObjectsFinal(session); err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjectsFinal: %s", err)
+	}
+	if len(obj) == 0 {
+		return 0, fmt.Errorf("no key found for the label: %s", key.label)
+	}
+	if len(obj) != 1 {
+		return 0, fmt.Errorf("got more than 1 key for the label: %s", key.label)
+	}
+
+	return obj[0], nil
+}
+
+func (c *Pkcs11Client) GetKeyMechanism(session pkcs11.SessionHandle, key pkcs11.ObjectHandle) (uint, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil),
+	}
+	attr, err := c.client.GetAttributeValue(session, pkcs11.ObjectHandle(key), template)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 GetAttributeValue: %s", err)
+	}
+
+	attrMap := GetAttributesMap(attr)
+	keyType := GetValueAsInt(attrMap[pkcs11.CKA_KEY_TYPE])
+
+	mechanism := uint(0)
+	switch keyType {
+	case pkcs11.CKK_AES:
+		if c.mechanism != 0 {
+			mechanism = c.mechanism
+		} else {
+			mechanism = pkcs11.CKM_AES_GCM
+		}
+	case pkcs11.CKK_RSA:
+		if c.mechanism != 0 {
+			mechanism = c.mechanism
+		} else {
+			mechanism = pkcs11.CKM_RSA_PKCS_OAEP
+		}
+	default:
+		return 0, fmt.Errorf("unsupported key type: %d", keyType)
+	}
+
+	return mechanism, nil
+}
+
+func (c *Pkcs11Client) GetMechAndPrefixSizeForKey(session pkcs11.SessionHandle, key pkcs11.ObjectHandle) (uint, int, error) {
+	mechanism, err := c.GetKeyMechanism(session, key)
+	if err != nil {
+		return 0, 0, err
+	}
+	switch mechanism {
+	case pkcs11.CKM_AES_GCM:
+		return mechanism, 16, nil
+	// Deprecated
+	case pkcs11.CKM_AES_CBC_PAD:
+		return mechanism, 16, nil
+	case pkcs11.CKM_AES_CBC:
+		return mechanism, 16, nil
+	// Consider others with no prefix
+	default:
+		return mechanism, 0, nil
+	}
+}
+
+func (c *Pkcs11Client) CreatePkcsParamsForKey(session pkcs11.SessionHandle, key pkcs11.ObjectHandle) ([]byte, *pkcs11.Mechanism, error) {
+	mechanismNum, prefixSize, err := c.GetMechAndPrefixSizeForKey(session, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Prefix is the IV or Nonce.
+	prefix, err := uuid.GenerateRandomBytes(prefixSize)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mechanism, err := c.GetPkcsParamsFromPrefix(session, key, mechanismNum, prefix)
+	if err != nil {
+		return nil, nil, err
+	}
+	return prefix, mechanism, nil
+}
+
+func (c *Pkcs11Client) GetPkcsParamsFromPrefix(session pkcs11.SessionHandle, key pkcs11.ObjectHandle, mechanism uint, prefix []byte) (*pkcs11.Mechanism, error) {
+	var params interface{}
+	switch mechanism {
+	case pkcs11.CKM_AES_GCM:
+		params = pkcs11.NewGCMParams(prefix, nil, 128)
```

Comment:
Hmm, check out this note: https://pkg.go.dev/github.com/miekg/pkcs11#NewGCMParams

> Note that some HSMs, like CloudHSM, will ignore the IV you pass in and write their own. As a result, to support all libraries, memory is not freed automatically, so that after the EncryptInit/Encrypt operation the HSM's IV can be read back out. It is up to the caller to ensure that Free() is called on the GCMParams object at an appropriate time, which is after Encrypt/Decrypt. As an example: 

.... so we likely want to make sure we save `params` here and update our copy of `prefix` from this value before doing the append in `Encrypt(...)`.